### PR TITLE
USB: Added VFS USB mass storage devices to the USB device list

### PR DESCRIPTION
**Pull request by brian218**
1. Implemented `control_transfer(0x80)`, which is used to retrieve the descriptor of a USB device.
2. Added VFS USB mass storage devices to the USB device list.

A number of System 357 games (e.g. Tekken and Razing Storm) retrieve the VID/PID/Serial Number information of a USB mass storage device via `control_transfer(0x80)` instead of  `sys_fs_fcntl(0xc000001c)`.

This PR makes those games pass the USB dongle check as long as proper USB information is set.

### CODE REVIEW

File: rpcs3/Emu/Io/usb_device.cpp
Comment (Florin9doi) [inline rpcs3/Emu/Io/usb_device.cpp:209]: buf_size may be bigger than device.data.size

File: rpcs3/Emu/Io/usb_device.cpp
Comment (Florin9doi) [inline rpcs3/Emu/Io/usb_device.cpp:270]:
```diff
- 	case 0x00:
+ 	case LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_STANDARD | LIBUSB_RECIPIENT_DEVICE: // 0x00
```

File: rpcs3/Emu/Io/usb_device.cpp
Comment (Florin9doi) [inline rpcs3/Emu/Io/usb_device.cpp:280]:
```diff
- 		case 0x00: get_status(false, false, buf, buf_size); break;
+ 		case LIBUSB_REQUEST_GET_STATUS: get_status(false, false, buf, buf_size); break;
```

File: rpcs3/Emu/Io/usb_device.cpp
Comment (Florin9doi) [inline rpcs3/Emu/Io/usb_device.cpp:277]:
```diff
-	case 0x80:
+ 	case LIBUSB_ENDPOINT_IN | LIBUSB_REQUEST_TYPE_STANDARD | LIBUSB_RECIPIENT_DEVICE: // 0x80
```

File: rpcs3/Emu/Io/usb_device.cpp
Comment (Florin9doi) [inline rpcs3/Emu/Io/usb_device.cpp:273]:
```diff
- 		case 0x09: usb_device::set_configuration(::narrow<u8>(wValue)); break;
+ 		case LIBUSB_REQUEST_SET_CONFIGURATION: usb_device::set_configuration(::narrow<u8>(wValue)); break;
```

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp:209]: I think buf_size can't be bigger than device.data.size().
Look at usb_device.h:124, the size of u8 data[0xFF] is fixed, and according to https://www.beyondlogic.org/usbnutshell/usb5.shtml the max reported size (bLength) is 255 (one byte).

Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: I still updated the code, however.

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Updated.

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Updated.

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Updated.

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Updated.

**Review by Florin9doi — APPROVED** [review]

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:2123]: This syntax is ugly. Don't return and assign something at the same time.
Also, why are you changing the return value here?

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:2197]: same here

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:2123]: I think VFS USB devices that don't have their path, vid, and pid set should be considered not mounted.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Updated.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Updated.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Alright. VID/PID's not being set should no longer bothers now.
A fallback value of 0x0000 is used instead.

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:300]: Can't you just add this as a function to the device_info struct?

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp]: Sure thing!

File: Utilities/Config.h
Comment (Megamouse) [inline Utilities/Config.h:647]:
```suggestion
		std::pair<u16, u16> get_hex_ids() const;
```

File: Utilities/Config.cpp
Comment (Megamouse) [inline Utilities/Config.cpp:562]:
```suggestion
std::pair<u16, u16> cfg::device_info::get_hex_ids() const
```

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:2119]: can stay const

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:2169]: can stay const

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:6]: obsolete

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:298]:
```suggestion
		const cfg::device_info device = g_cfg_vfs.get_device(g_cfg_vfs.dev_usb, fmt::format("/dev_usb%03d", i));
```

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:2129]: I think this file now has an unused `#include <charconv>` that can be removed

File: Utilities/Config.h
Comment (brian218) [reply Utilities/Config.h]: Aha, thanks a lot for the wise advice!

File: Utilities/Config.cpp
Comment (brian218) [reply Utilities/Config.cpp]: Modified. Thanks!

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Reverted.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp]: Reverted.

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp]: Removed.

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp]: Reverted.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:2129]: Removed.

**Comment by brian218** [comment]
I have tested this PR with some System 357 games, which either use control_transfer(0x80) or sys_fs_fcntl(0xc0000015&0xc000001c) for USB dongle check, and everything worked well.

**Comment by brian218** [comment]
SCEEXE000_screenshot_2023_01_15

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:997]: missing default

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:1009]: missing default

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp:997]: Added.

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp:1010]: Added.

**Comment by brian218** [comment]
> Please stop requesting reviews from me ;) I've removed myself as reviewer 3 times now for a reason.

@Megamouse You are one of the persons who have rights to make decisions about a PR, and you are one of the two persons whom I am able to request review from.
I'm just wondering what _reason_ is preventing this PR from being accepted for days.

**Comment by Megamouse** [comment]
Reason 1: We all have jobs and RL etc., so you just gotta wait until someone has time or takes interest.
Reason 2: You keep changing stuff daily (be it due to reviews or not), so people have to check it again after each change.
Reason 3: I'm not the expert in usb stuff, so I won't merge it anyway before I got some second opinion.

**Comment by brian218** [comment]
> Reason 1: We all have jobs and RL etc., so you just gotta wait until someone has time or takes interest. Reason 2: You keep changing stuff daily (be it due to reviews or not), so people have to check it again after each change. Reason 3: I'm not the expert in usb stuff, so I won't merge it anyway before I got some second opinion.

@Megamouse Thanks for the clarification.
2. It can be due to "Update with rebase" option to keep up with the master branch (I prefer using this to using "Update with merge commit"), since the master branch is often updated several times a day.
Image
3. @Florin9doi, who has already had a number of USB-related PRs merged into RPCS3, has done some reviews regarding the usb_device stuff.

**Comment by RipleyTom** [comment]
You do not need to update your branch, as long as there is no conflict it'll be rebased automatically before getting merged.

File: rpcs3/Emu/Io/usb_device.cpp
Comment (RipleyTom) [inline rpcs3/Emu/Io/usb_device.cpp:242]: le_t<u16>

Comment (RipleyTom) [inline rpcs3/Emu/Io/usb_device.cpp:242]: le_t<u16>

**Review by RipleyTom — CHANGES_REQUESTED** [review]
Apart from that lgtm

File: rpcs3/Emu/Io/usb_device.cpp
Comment (brian218) [reply rpcs3/Emu/Io/usb_device.cpp]: Done.

**Review by RipleyTom — APPROVED** [review]

**Pull request merged**
